okhttp/src/main/java/okhttp3/Headers.kt

```diff
@@ -0,0 +1,410 @@
+          addLenient(line.substring(0, index), line.substring(index + 1))
+        }
+        line.startsWith(":") -> {
+          // Work around empty header names and header names that start with a
+          // colon (created by old broken SPDY versions of the response cache).
+          addLenient("", line.substring(1)) // Empty header name.
+        }
+        else -> {
+          // No header name.
+          addLenient("", line)
+        }
+      }
+    }
+
+    /** Add an header line containing a field name, a literal colon, and a value. */
+    fun add(line: String): Builder {
+      val index = line.indexOf(":")
+      require(index != -1) { "Unexpected header: $line" }
+      return add(line.substring(0, index).trim { it <= ' ' }, line.substring(index + 1))
+    }
+
+    /**
+     * Add a header with the specified name and value. Does validation of header names and values.
+     */
+    fun add(name: String, value: String): Builder {
+      checkName(name)
+      checkValue(value, name)
+      return addLenient(name, value)
+    }
+
+    /**
+     * Add a header with the specified name and value. Does validation of header names, allowing
+     * non-ASCII values.
+     */
+    fun addUnsafeNonAscii(name: String, value: String): Builder {
+      checkName(name)
+      return addLenient(name, value)
+    }
+
+    /**
+     * Adds all headers from an existing collection.
+     */
+    fun addAll(headers: Headers) = apply {
+      for (i in 0 until headers.size()) {
+        addLenient(headers.name(i), headers.value(i))
+      }
+    }
+
+    /**
+     * Add a header with the specified name and formatted date. Does validation of header names and
+     * value.
+     */
+    fun add(name: String, value: Date) = apply {
+      add(name, HttpDate.format(value))
+    }
+
+    /**
+     * Add a header with the specified name and formatted instant. Does validation of header names
+     * and value.
+     */
+    @IgnoreJRERequirement
+    fun add(name: String, value: Instant): Builder {
+      return add(name, Date(value.toEpochMilli()))
+    }
+
+    /**
+     * Set a field with the specified date. If the field is not found, it is added. If the field is
+     * found, the existing values are replaced.
+     */
+    operator fun set(name: String, value: Date) = apply {
+      set(name, HttpDate.format(value))
+    }
+
+    /**
+     * Set a field with the specified instant. If the field is not found, it is added. If the field
+     * is found, the existing values are replaced.
+     */
+    @IgnoreJRERequirement
+    operator fun set(name: String, value: Instant) = apply {
+      return set(name, Date(value.toEpochMilli()))
+    }
+
+    /**
+     * Add a field with the specified value without any validation. Only appropriate for headers
+     * from the remote peer or cache.
+     */
+    internal fun addLenient(name: String, value: String) = apply {
+      namesAndValues.add(name)
+      namesAndValues.add(value.trim { it <= ' ' })
```

Comment:
Feels like we probably want to pull this `trim { .. }` out to a custom extension method instead of repeating it. Can be follow-up.